internalshared/configutil/datadog_sink.go

```diff
@@ -0,0 +1,140 @@
+package configutil
+
+import (
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/armon/go-metrics/datadog"
+	"github.com/mitchellh/cli"
+)
+
+type DatadogSink struct {
+	tags                 []string
+	addr                 string
+	hostName             string
+	propagateHostname    bool
+	sink                 *datadog.DogStatsdSink
+	logger               cli.Ui
+	attemptedToConnectAt *time.Time
+	lock                 sync.Mutex
+}
+
+func NewDatadogSink(addr string, hostName string, logger cli.Ui) *DatadogSink {
+	return &DatadogSink{
+		addr:              addr,
+		hostName:          hostName,
+		propagateHostname: false,
+		logger:            logger,
+	}
+}
+
+func (s *DatadogSink) SetTags(tags []string) {
+	s.lock.Lock()
+	s.tags = tags
+	s.lock.Unlock()
+
+	sink := s.getSink()
+	if sink == nil {
+		return
+	}
+	sink.SetTags(tags)
+}
+
+func (s *DatadogSink) EnableHostNamePropagation() {
+	s.lock.Lock()
+	s.propagateHostname = true
+	s.lock.Unlock()
+
+	sink := s.getSink()
+	if sink == nil {
+		return
+	}
+	sink.EnableHostNamePropagation()
+}
+
+// Implementation of methods in the MetricSink interface
+
+func (s *DatadogSink) SetGauge(key []string, val float32) {
+	s.SetGaugeWithLabels(key, val, nil)
+}
+
+func (s *DatadogSink) IncrCounter(key []string, val float32) {
+	s.IncrCounterWithLabels(key, val, nil)
+}
+
+func (s *DatadogSink) EmitKey(key []string, val float32) {
+	sink := s.getSink()
+	if sink == nil {
+		return
+	}
+	sink.EmitKey(key, val)
+}
+
+func (s *DatadogSink) AddSample(key []string, val float32) {
+	s.AddSampleWithLabels(key, val, nil)
+}
+
+func (s *DatadogSink) getSink() *datadog.DogStatsdSink {
+	s.lock.Lock()
```

Comment:
I think at this point the change is "correct", in that we could merge it now without causing breakage.  I'm a little concerned about the performance implications here though.  Since a basic mutex can only be locked by a single goroutine, we're now serializing all metrics writers when a datadog sink exists.  Maybe we're better off using an RWMutex.  Then we'd only Rlock in the happy path, escalating to a write lock if we get to the point where we need to actually modify s.attemptedToConnectAt.

Reply:
@ncabatoff Done. 
I refactored the code to make it testable, because the code gets complicated.